### ISSUE TITLE
gencli: issue with oneof with matching name

### ISSUE THREAD

**Issue by tmc** [googleapis/gapic-generator-go#368]
I'm not 100% sure how to explain the cause of this bug but this message:
https://github.com/googleapis/googleapis/blob/master/google/cloud/secretmanager/v1/resources.proto#L122

Doesn't produce code that compiles: https://github.com/tmc/secretsctl/runs/616003783#step:5:5

**Comment by noahdietz** [comment]
Copying the error messages here:

```
# github.com/tmc/secretsctl/cmd/secretsctl
cmd/secretsctl/update-secret.go:40:57: cannot use new(secretmanager.Replication_Automatic_) (type *secretmanager.Replication_Automatic_) as type secretmanager.Replication_Automatic in assignment
cmd/secretsctl/update-secret.go:42:59: cannot use new(secretmanager.Replication_UserManaged) (type *secretmanager.Replication_UserManaged) as type secretmanager.Replication_UserManaged in assignment
cmd/secretsctl/update-secret.go:91:54: cannot use &UpdateSecretInputSecretReplicationReplicationAutomatic (type *secretmanager.Replication_Automatic) as type secretmanager.isReplication_Replication in assignment:
	*secretmanager.Replication_Automatic does not implement secretmanager.isReplication_Replication (missing secretmanager.isReplication_Replication method)
cmd/secretsctl/update-secret.go:94:54: cannot use &UpdateSecretInputSecretReplicationReplicationUserManaged (type *secretmanager.Replication_UserManaged) as type secretmanager.isReplication_Replication in assignment:
	*secretmanager.Replication_UserManaged does not implement secretmanager.isReplication_Replication (missing secretmanager.isReplication_Replication method)
cmd/secretsctl/update-secret.go:110:60: UpdateSecretInputSecretReplicationReplicationUserManaged.UserManaged undefined (type secretmanager.Replication_UserManaged has no field or method UserManaged)
cmd/secretsctl/update-secret.go:115:11: undefined: secretmanager.Secret_LabelsEntry
cmd/secretsctl/update-secret.go:121:44: first argument to append must be slice; have map[string]string
```

**Comment by noahdietz** [comment]
Thanks @tmc I will take a look.

**Comment by noahdietz** [comment]
Ok so the Replication_Automatic and Replication_UserManaged errors are because the oneof selector type (that implements the generated oneof interface) is named `Replication_Automatic_` and `Replication_UserManaged_` (with the trailing `_`). `go_cli` didn't account for that possibility and I believe splits on `_` to construct variable names. I believe the generated protobuf code has these trailing `_` because the containing message is `Replication`, with _nested_ messages `Automatic` and `UserManaged`, a oneof selector named `replication` and the oneof fields named `automatic` and `user_managed`. 

In the generated `go_cli`, changing the oneof selector variables like so:

```
// create-secret.go
var CreateSecretInputSecretReplicationReplicationAutomatic secretmanagerpb.Replication_Automatic_

var CreateSecretInputSecretReplicationReplicationUserManaged secretmanagerpb.Replication_UserManaged_

// update-secret.go
var UpdateSecretInputSecretReplicationReplicationAutomatic secretmanagerpb.Replication_Automatic_

var UpdateSecretInputSecretReplicationReplicationUserManaged secretmanagerpb.Replication_UserManaged_
```

resolves the related compilation errors and leaves me with the following errors (which I have yet to investigate):
```
./create-secret.go:117:11: undefined: "google.golang.org/genproto/googleapis/cloud/secretmanager/v1".Secret_LabelsEntry
./create-secret.go:123:44: first argument to append must be slice; have map[string]string
./update-secret.go:115:11: undefined: "google.golang.org/genproto/googleapis/cloud/secretmanager/v1".Secret_LabelsEntry
./update-secret.go:121:44: first argument to append must be slice; have map[string]string
```

**Comment by noahdietz** [comment]
For fun, [this LoC](https://github.com/golang/protobuf/blob/master/protoc-gen-go/generator/generator.go#L2316) seems to be what appends the `_` suffix when there are naming collisions.

**Comment by noahdietz** [comment]
The `undefined: secretmanager.Secret_LabelsEntry` issue is a mishandling of the `map` type. Because protobuf represents maps as ([source](https://developers.google.com/protocol-buffers/docs/proto3#maps)): 

```proto
message MapFieldEntry {
  key_type key = 1;
  value_type value = 2;
}

repeated MapFieldEntry map_field = N;
```

`go_cli` mishandles it, seeing it as a repeated message with a nested definition hence the `Secret_LabelsEntry`. Seeing what I can do...

**Comment by noahdietz** [comment]
Quick update: really sorry @tmc a bunch of stuff popped up in the last week and this got bumped off my plate. Still busy atm, idk when I'll be able get to this

**Comment by tmc** [comment]
> Quick update: really sorry @tmc a bunch of stuff popped up in the last week and this got bumped off my plate. Still busy atm, idk when I'll be able get to this

Thanks for the update.

**Comment by noahdietz** [comment]
@tmc apologies for the delay on this. I've merged two fixes, one that repairs the oneof handling to avoid collisions and another to support `map<string, string>` fields via a repeated `key=value` flag. I was able to generate and compile the `secretmanager` CLI. I released the changes in [v0.14.2](https://github.com/googleapis/gapic-generator-go/releases/v0.14.2). Feel free to reopen if it doesn't work out.

**Comment by tmc** [comment]
> @tmc apologies for the delay on this. I've merged two fixes, one that repairs the oneof handling to avoid collisions and another to support `map<string, string>` fields via a repeated `key=value` flag. I was able to generate and compile the `secretmanager` CLI. I released the changes in [v0.14.2](https://github.com/googleapis/gapic-generator-go/releases/v0.14.2). Feel free to reopen if it doesn't work out.

Wonderful! Looks like it's working out. Thanks.
